### PR TITLE
Implement LSP Text Document Synchronization

### PR DESCRIPTION
**Pull request by schrieveslaach**
# Description

This commit provides the initial implementation of text document synchronization of the language server protocol: textDocument/didOpen, textDocument/didChange and textDocument/didClose (see #10794)

# User-Facing Changes

Completions work better then before (see [comment here](https://github.com/nushell/vscode-nushell-lang/pull/162#issuecomment-1791036239)). The demo show that Nushell opens the line buffer editor, that starts Neovim with Nushell as configured language server which has completions working.

output

# Tests + Formatting

# After Submitting

### CODE REVIEW

**Comment by schrieveslaach** [comment]
@fdncred, improve completions are working now better. :wink:

**Comment by fdncred** [comment]
whoa! that looks great! looks like clippy is objecting to something, as usual :)

File: crates/nu-lsp/src/notification.rs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        if let Ok(file_path) = params.text_document.uri.to_file_path() {
+            for content_change in params.content_changes.into_iter() {
+                let entry = self.ropes.entry(file_path.clone());
+                match (content_change.range, content_change.range) {
+                    (Some(range), _) => {
+                        entry.and_modify(|rope| {
+                            let start = Self::lsp_position_to_location(&range.start, rope);
+                            let end = Self::lsp_position_to_location(&range.end, rope);
+
+                            println!("{rope:?}");
```
Comment (fdncred) [inline crates/nu-lsp/src/notification.rs:63]: seems like these debug println!'s might not need to be here?
```suggestion
```

File: crates/nu-lsp/src/notification.rs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+                            rope.remove(start..end);
+                            println!("{rope:?}");
```
Comment (fdncred) [inline crates/nu-lsp/src/notification.rs:65]:
```suggestion
```

File: crates/nu-lsp/src/notification.rs [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                            rope.insert(start, &content_change.text);
+                            println!("{rope:?}");
```
Comment (fdncred) [inline crates/nu-lsp/src/notification.rs:67]:
```suggestion
```

**Comment by fdncred** [comment]
I'm seeing weird stuff in helix. I can't get hovers to work, and completions only seem to work when it's in the root of a script and not inside of a def.
image

After about 20 seconds waiting for a hover, this appears.
image

and when I exit it takes a while and then it finally exits with this message.
image

**Comment by fdncred** [comment]
Hmmm, it's not really working in vscode either. goto-definition isn't working, nor is hover.
```
[Error - 4:05:36 PM] Request textDocument/hover failed.
Error: The received response has neither a result nor an error property.
    at handleInvalidMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:672:40)
    at handleMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:348:13)
    at processMessageQueue (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:362:17)
    at Immediate.<anonymous> (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:334:13)
    at processImmediate (node:internal/timers:476:21)
    at process.callbackTrampoline (node:internal/async_hooks:130:17)
[Error - 4:05:51 PM] Request textDocument/hover failed.
Error: The received response has neither a result nor an error property.
    at handleInvalidMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:672:40)
    at handleMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:348:13)
    at processMessageQueue (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:362:17)
    at Immediate.<anonymous> (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:334:13)
    at processImmediate (node:internal/timers:476:21)
    at process.callbackTrampoline (node:internal/async_hooks:130:17)
[Error - 4:05:58 PM] Request textDocument/hover failed.
Error: The received response has neither a result nor an error property.
    at handleInvalidMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:672:40)
    at handleMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:348:13)
    at processMessageQueue (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:362:17)
    at Immediate.<anonymous> (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:334:13)
    at processImmediate (node:internal/timers:476:21)
    at process.callbackTrampoline (node:internal/async_hooks:130:17)
[Error - 4:06:14 PM] Request textDocument/hover failed.
Error: The received response has neither a result nor an error property.
    at handleInvalidMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:672:40)
    at handleMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:348:13)
    at processMessageQueue (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:362:17)
    at Immediate.<anonymous> (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:334:13)
    at processImmediate (node:internal/timers:476:21)
    at process.callbackTrampoline (node:internal/async_hooks:130:17)
[Error - 4:06:22 PM] Request textDocument/hover failed.
Error: The received response has neither a result nor an error property.
    at handleInvalidMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:672:40)
    at handleMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:348:13)
    at processMessageQueue (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:362:17)
    at Immediate.<anonymous> (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:334:13)
    at processImmediate (node:internal/timers:476:21)
    at process.callbackTrampoline (node:internal/async_hooks:130:17)
[Error - 4:06:22 PM] Request textDocument/hover failed.
Error: The received response has neither a result nor an error property.
    at handleInvalidMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:672:40)
    at handleMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:348:13)
    at processMessageQueue (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:362:17)
    at Immediate.<anonymous> (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:334:13)
    at processImmediate (node:internal/timers:476:21)
    at process.callbackTrampoline (node:internal/async_hooks:130:17)
[Error - 4:06:27 PM] Request textDocument/hover failed.
Error: The received response has neither a result nor an error property.
    at handleInvalidMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:672:40)
    at handleMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:348:13)
    at processMessageQueue (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:362:17)
    at Immediate.<anonymous> (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:334:13)
    at processImmediate (node:internal/timers:476:21)
    at process.callbackTrampoline (node:internal/async_hooks:130:17)
[Error - 4:06:27 PM] Request textDocument/hover failed.
Error: The received response has neither a result nor an error property.
    at handleInvalidMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:672:40)
    at handleMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:348:13)
    at processMessageQueue (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:362:17)
    at Immediate.<anonymous> (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:334:13)
    at processImmediate (node:internal/timers:476:21)
    at process.callbackTrampoline (node:internal/async_hooks:130:17)
[Error - 4:06:30 PM] Request textDocument/definition failed.
Error: The received response has neither a result nor an error property.
    at handleInvalidMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:672:40)
    at handleMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:348:13)
    at processMessageQueue (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:362:17)
    at Immediate.<anonymous> (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:334:13)
    at processImmediate (node:internal/timers:476:21)
    at process.callbackTrampoline (node:internal/async_hooks:130:17)
[Error - 4:06:30 PM] Request textDocument/hover failed.
Error: The received response has neither a result nor an error property.
    at handleInvalidMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:672:40)
    at handleMessage (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:348:13)
    at processMessageQueue (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:362:17)
    at Immediate.<anonymous> (/Users/fdncred/src/vscode-nushell-lang/client/node_modules/vscode-jsonrpc/lib/common/connection.js:334:13)
    at processImmediate (node:internal/timers:476:21)
    at process.callbackTrampoline (node:internal/async_hooks:130:17)

```

**Comment by schrieveslaach** [comment]
@fdncred, are these errors popping up with this PR or are they are also popping up on main?

**Comment by fdncred** [comment]
@schrieveslaach There's specifically with this PR.

**Comment by schrieveslaach** [comment]
@fdncred, could you give it another try?

**Comment by fdncred** [comment]
I'm still getting this on the latest version
```
[Error - 10:48:35 AM] Request textDocument/hover failed.
Error: The received response has neither a result nor an error property.
    at handleInvalidMessage (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:672:40)
    at handleMessage (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:348:13)
    at processMessageQueue (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:362:17)
    at Immediate.<anonymous> (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:334:13)
    at processImmediate (node:internal/timers:476:21)
    at process.callbackTrampoline (node:internal/async_hooks:130:17)
[Error - 10:48:36 AM] Request textDocument/hover failed.
Error: The received response has neither a result nor an error property.
    at handleInvalidMessage (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:672:40)
    at handleMessage (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:348:13)
    at processMessageQueue (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:362:17)
    at Immediate.<anonymous> (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:334:13)
    at processImmediate (node:internal/timers:476:21)
    at process.callbackTrampoline (node:internal/async_hooks:130:17)
[Error - 10:48:36 AM] Request textDocument/hover failed.
Error: The received response has neither a result nor an error property.
    at handleInvalidMessage (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:672:40)
    at handleMessage (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:348:13)
    at processMessageQueue (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:362:17)
    at Immediate.<anonymous> (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:334:13)
    at processImmediate (node:internal/timers:476:21)
    at process.callbackTrampoline (node:internal/async_hooks:130:17)
[Error - 10:48:37 AM] Request textDocument/hover failed.
Error: The received response has neither a result nor an error property.
    at handleInvalidMessage (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:672:40)
    at handleMessage (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:348:13)
    at processMessageQueue (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:362:17)
    at Immediate.<anonymous> (d:\vscode-nushell-lang\client\node_modules\vscode-jsonrpc\lib\common\connection.js:334:13)
    at processImmediate (node:internal/timers:476:21)
    at process.callbackTrampoline (node:internal/async_hooks:130:17)
```

**Comment by fdncred** [comment]
On the server side, I got this. Just saw it.
image

**Comment by schrieveslaach** [comment]
That was a good hint and made me look into the spec:

>  This means for example that a null value has to be explicitly listed and that a mandatory property must be listed even if a falsify value might exist. Based on the basic JSON structures, the actual requests with their responses and the notifications are described.

So, in case that there is no result, `null` has to be serialized and following handling has to be done. 

```rust
                   param_handler(engine_state, &params)
                        .and_then(|response| serde_json::to_value(response).ok())
                        .unwrap_or(serde_json::Value::Null),
```

**Comment by schrieveslaach** [comment]
The duplication issue makes me think if you have two LSPs running in the background… I can't reproduce it in Neovim. Do you have the same issue in Helix?

**Comment by fdncred** [comment]
In helix the inlays look good. The completions seem good. The hovers seem good (exept for markdown formatting). I can't seem to get goto definition working.

`g+D`
image
`g+d`
image

**Comment by fdncred** [comment]
ugh, i take it back. my language server is not set to nu --lsp

**Comment by fdncred** [comment]
ok, i updated my languages.toml to support `nu --lsp`. I kind of get the same results as I mentioned here https://github.com/nushell/nushell/pull/10941#issuecomment-1802645325 - hard to tell if it's working or if it's the tree-sitter grammar doing things. I just don't understand enough about helix.

**Comment by schrieveslaach** [comment]
Strange. I'll try to test the LSP in Helix later. Maybe @jokeyrhyme could test this PR too in Vscode. 

Meanwhile, I implemented diagnostics:

output

**Comment by fdncred** [comment]
That's so cool!!!!

**Comment by fdncred** [comment]
strangely enough, goto-definition on macos with helix is working.

**Comment by schrieveslaach** [comment]
> strangely enough, goto-definition on macos with helix is working, as well as hover, but not completions.

Are you testing on Windows where it is not working?

**Comment by NotLebedev** [comment]
Tried running it on Linux (nixOS) with helix. Currently no luck.
My `languages.toml` looks like:
```toml
[[language]]
name = "nix"
rulers = [100]

[language.formatter]
command = "/nix/store/iqxdvd5gyckvs6k9al85rgdgqxc6sx0m-nixpkgs-fmt-1.3.0/bin/nixpkgs-fmt"

[[language]]
language-servers = ["nuls"]
name = "nu"

[language-server.nuls]
args = ["--lsp"]
command = "/nix/store/ib059h1skbr272kbn80sig1yk26466za-nu-0.86.1/bin/nu"
```
When I do lsp-restart I get "Language server exited" and
```
2023-11-09T12:13:02.672 helix_lsp::transport [ERROR] nuls err: <- StreamClosed
```
in log.

**Comment by fdncred** [comment]
@NotLebedev This is what I added to languages.toml. Note that I'm using the latest master branch in helix.
on line 50 under `[language-server]`
```toml
nu-lsp = { command = "nu", args = [ "--lsp" ] }
```
and later on, line 1858
```toml
[[language]]
name = "nu"
scope = "source.nu"
injection-regex = "nu"
file-types = ["nu"]
shebangs = ["nu"]
roots = []
comment-token = "#"
indent = { tab-width = 2, unit = "  " }
language-servers = [ "nu-lsp" ]
```

**Comment by fdncred** [comment]
> Are you testing on Windows where it is not working?

Yes, I was testing in Windows with helix and vscode, and the same on MacOS. Windows was behaving oddly. I'm going to try again today on Windows.

**Comment by NotLebedev** [comment]
> latest master branch

Oh, that's the deal. Thanks. I was on 23.05 and upgraded to 23.10 and that was enough.

It seems like everything is working (goto defentinion/declaration included) except for syntax highlights it seems.
For example look at how `$nu.home-path` is colored in helix:
image
vs command line
image

**Comment by fdncred** [comment]
> except syntax highlights it seems.

syntax highlight is controlled by the tree-sitter grammar right now, no the lsp.

**Comment by NotLebedev** [comment]
> syntax highlight is controlled by the tree-sitter grammar right now, no the lsp.

Helix still has this limitation 🤔 I thought they discussed adding syntax highlighting back in January. Well than for me and my scripts everything works fine. Great PR! Great work on this feature!

**Comment by fdncred** [comment]
@NotLebedev you can see here what's currently implemented https://github.com/nushell/nushell/issues/10794 - I think there may also be diagnostics that are implemented, not sure, i haven't tried to trigger them yet.

**Comment by fdncred** [comment]
Testing on Windows again.

### Hover over custom command
image

### Hover over built-in command
image

### Hover over variable
image

### Goto-definition of custom command `g` then `d` - Not Working in helix.

### Completions seem to be working now. I can't get them to work in the latest main branch but they're working in this PR 🎉 
image

### Inlays
I think these are from tree-sitter

### Diagnostics
I tried the example in the gif but couldn't get it to work right. I typed the function as in the gif, no error message. then I hit undo and got it

A few more steps
image
No error
image
After undo
image

I can't get it to do this every time. It's weird.

**Comment by fdncred** [comment]
I'm so sick of messing this up. I had `nuls` installed and it was working vs `nu --lsp`.

I must have something misconfigured.
image

**Comment by fdncred** [comment]
I got goto-def and hover working in MacOS today with nuls removed from my file system to make sure it's not being used. I can't get the diagnostics to work nor completions.

I got this while trying to get completions to work.
image

**Comment by jokeyrhyme** [comment]
From the teensy bit of testing I've done so far, I _think_ `nu --lsp` does not respond to Ctrl+C when running separately, but it's always shutdown when helix sends it the proper messages

I think helix occasionally runs into that shutdown timeout with most LSPs, so that part might be a helix thing (or even an issue with every editor, just that helix actually warns you)

**Comment by schrieveslaach** [comment]
I raised a PR against the LSP server crate to make sure that CTRL + C can be handled properly (see rust-lang/rust-analyzer#15894) and updated this PR.

**Comment by schrieveslaach** [comment]
output

I tried the LSP with Helix on Linux. Seems to work.

**Comment by fdncred** [comment]
I just compiled with the latest changes and it's working in helix for me now too!
image

Just to be clear, I have this in my ~/.config/helix/config.toml
```toml
theme = "dark_plus"

[editor]
mouse = true
true-color = true
#rulers = [80,100,120]
auto-info = true
#auto-pairs = true
gutters = ["diff", "diagnostics", "line-numbers", "spacer"]

[editor.cursor-shape]
insert = "bar"
normal = "block"
select = "underline"

[editor.file-picker]
hidden = false

[editor.lsp]
enable = true #Enables LSP integration. Setting to false will completely disable language servers regardless of language settings.
display-messages = true #Display LSP progress messages below statusline1
auto-signature-help = true #Enable automatic popup of signature help (parameter hints)
display-inlay-hints = true #Display inlay hints2
display-signature-help-docs = true #Display docs under signature help popup
snippets = true #Enables snippet completions. Requires a server restart (:lsp-restart) to take effect after :config-reload/:set.
goto-reference-include-declaration = true #Include declaration in the goto references popup.

[editor.auto-pairs]
'(' = ')'
'{' = '}'
'[' = ']'
'"' = '"'
'`' = '`'
'<' = '>'

```

Then this in the languages.toml
```toml
[language-server]
... a bunch of other language servers
nu-lsp = { command = "nu", args = [ "--lsp" ] }

... lots of other stuff
[[language]]
name = "nu"
scope = "source.nu"
injection-regex = "nu"
file-types = ["nu"]
shebangs = ["nu"]
roots = []
comment-token = "#"
indent = { tab-width = 2, unit = "  " }
language-servers = [ "nu-lsp" ]
```

**Comment by fdncred** [comment]
We usually have a pause to landing PRs a few days before a release and a few days after a release. I'd like to land this as soon as we resume landing again.

**Comment by sholderbach** [comment]
@fdncred `main` branch is free for you to do the honors 

Thanks for the quick iteration on this @schrieveslaach

**Comment by fdncred** [comment]
Let's go!

**Comment by schrieveslaach** [comment]
Thanks for merging it. :heart_eyes:

**Comment by fdncred** [comment]
> Thanks for merging it. 😍

We're happy to keep chipping away at this thing. Keep up the good work!

**Comment by fdncred** [comment]
The one thing that concerns me about this PR is that we have a git dependency. I'm just hoping the rust-analyzer folks will land that PR soon, because if it doesn't land by our next release (Dec 12th), we will have to revert this PR.

**Comment by jokeyrhyme** [comment]
> The one thing that concerns me about this PR is that we have a git dependency. I'm just hoping the rust-analyzer folks will land that PR soon, because if it doesn't land by our next release (Dec 12th), we will have to revert this PR.

That was specifically for reacting to Ctrl+C, which I'd suggest probably isn't super critical and could be selectively reverted/removed

**Comment by schrieveslaach** [comment]
@jokeyrhyme, @fdncred, https://github.com/rust-lang/rust-analyzer/pull/15894 has been merged. Do you want me to update the main branch pointing to the merged version? No release yet on lsp-server.

**Comment by fdncred** [comment]
@schrieveslaach what's the latest on this git dependency? has there been a release that includes it? if so, can we get it in?

**Comment by schrieveslaach** [comment]
@fdncred, no release yet… However, I created #11250 to downgrade to the latest release version.

**Comment by fdncred** [comment]
@schrieveslaach Thanks. Our next release is schedule for Dec 12. So we may have to land that PR. I appreciate you creating it.

**Comment by schrieveslaach** [comment]
For anyone listening #11252 upgrades to the latest release of lsp-server including the ctrl-c support

**Pull request merged**